src/v/raft/coordinated_recovery_throttle.cc

```diff
@@ -185,73 +219,128 @@ ss::future<> coordinated_recovery_throttle::do_coordinate_tick() {
     }
 
-    // Any updates to the rate are picked up in the next tick.
-    auto total_rate = _rate_binding();
-    auto fair_shard_rate = fair_rate_per_shard();
+    // Any updates to the rate are picked up here in the beginning of the tick.
+    // Clamp up for each shard to receive at least one unit.
+    // Clamp down to avoid overflow.
+    // Retain zero as it is to block all recovery (used in tests).
+    ssize_t total_rate = _rate_binding()
+                           ? std::clamp(
+                               _rate_binding(),
+                               size_t{ss::smp::count},
+                               std::numeric_limits<size_t>::max() / 4)
+                           : 0;
+
+    // Fetch the requirements from all shards.
+    const auto capacity_requirements = co_await container().map(
+      std::mem_fn(&coordinated_recovery_throttle::required_capacity));
 
-    if (unlikely(_use_static_allocation())) {
-        co_return co_await reset_capacity_all_shards(fair_shard_rate);
+    // Remaining units if positive, overused if negative
+    ssize_t total_remaining_units = 0;
+    for (const auto& req : capacity_requirements) {
+        total_remaining_units += req.remaining_units;
+        vlog(
+          raftlog.trace,
+          "Shard capacity request: remaining_units={}, "
+          "desired_additional_capacity={}",
+          req.remaining_units,
+          req.desired_additional_capacity);
     }
 
-    // Step 1: Fetch the required capacity from all shards.
-    const auto capacity_requirements = co_await container().map(
-      [](coordinated_recovery_throttle& crt) {
-          return crt.required_capacity();
-      });
+    ssize_t unused_units = std::max(ssize_t{0}, total_remaining_units);
+    // Do not carry over to the next tick. What is unused is lost.
+    // Can be negative if config changed to a lower rate
+    ssize_t to_add = total_rate - unused_units;
 
-    // Step 2: Compute what unused bandwidth each shard has to offer.
-    // Here we guarantee that each shard is assigned atleast fair_shard_rate
-    // if needed. Anything beyond that is unused/deficit.
-    size_t total_bandwidth_unused = 0;
-    size_t total_bandwidth_deficit = 0;
-
-    for (auto req : capacity_requirements) {
-        if (req <= fair_shard_rate) {
-            total_bandwidth_unused += (fair_shard_rate - req);
-        } else {
-            total_bandwidth_deficit += (req - fair_shard_rate);
-        }
+    if (unlikely(_use_static_allocation())) {
+        // Don't take individual shard requests into account, distribute equally
+        auto fair_shard_rate = to_add / ss::smp::count;
+        co_return co_await renew_capacity_all_shards(fair_shard_rate);
     }
 
+    // Surplus after we satisfy hard requirements (that is, bring all shard
+    // rates to 0 assuming they stay as they are while we're calculating).
+    // Negative means net overuse.
+    ssize_t avg_surplus = (to_add + total_remaining_units) / ss::smp::count;
```

Comment:
is there any behavioral change in this redistribution logic? I'm not really sure from the diff itself but just wanted to double check.